enhancements/declarative-index-config.md

```diff
@@ -563,7 +537,7 @@ etcdoperator.v0.9.4-clusterwide  quay.io/operatorhubio/etcd:v0.9.4-clusterwide
 etcdoperator.v0.9.4              quay.io/operatorhubio/etcd:v0.9.4              0.9.4                                etcdoperator.v0.9.2   
 ```
 
-This information will be captured in the `properties.<"olm.channel">.replaces`, `properties.skips` and `properties.skipRange` of the `olm.bundle` blob. In the `properties.<"olm.channel">` object for each channel, the bundle that replaces a previous bundle is captured in the `replaces` field. An upgrade edge in a channel is represented by: bundle.properties<"olm.channel">.replaces <- bundle name. 
+This information will be captured in the `properties.<"olm.channel">.values.replaces`, `properties.skips` and `properties.skipRange` of the `olm.bundle` blob. In the `properties.<"olm.channel">` object for each channel, the bundle that replaces a previous bundle is captured in the `values.replaces` field. An upgrade edge in a channel is represented by: `bundle.properties.<"olm.channel">.values.replaces` <- `bundle.name`.
```

Comment:
> An upgrade edge in a channel is represented by: `bundle.properties.<"olm.channel">.values.replaces` <- `bundle.name`.

It might be helpful to describe this in more detail or provide a link to existing docs. This text makes it seem like `replaces` edges are the ONLY edges used to build the graph and determine a channel head.

But other areas of the doc say that `skips` and `skipRange` are also important and that they may have an impact on how the channel_entry table is built and thus how the upgrade graph looks, which then translates to semantics that will need to be handled when building the declarative config representation of the same data.

In some of my testing, it seemed like `skips` is accounted for in the `channel_entry` table, and it is unclear to me if `skipRange` is. Until I accounted for `skips` edges, I was noticing multiple bundles with no incoming edges. i.e. something like this with `b1` and `b3`:
```
          b3 --> b5
                /  \
b1 --> b2 --> b4 --> b6 
```

After accounting for `skips` the graph turned into:
```
          b3 --> b5
         /      /  \
b1 --> b2 --> b4 --> b6 
```

Since `replaces` allows only one outgoing edge from a node, it seems necessary to include `skips` and `skipRange` as key elements for determining the final graph.

Reply:
> But other areas of the doc say that skips...

You are correct about skips. It was created to allow users to omit an operator version from their graph, while still providing an edge from it.

> In some of my testing, it seemed like skips is accounted for in the channel_entry table, and it is unclear to me if skipRange is.

`skipRange` is not involved in the `channel_entry` table, but it is used at runtime. It effectively adds edges from the head of particular channel to any _installed_ operator within the specified range.

Reply:
Ok yeah, this is coming back to me. Sorry for having to have this explained multiple times :)

So I think what I'm hearing is that to determine the channel head from a set of bundles, I can count up incoming edges from skips and replaces and the bundle that does not have any incoming edges is the head. And there MUST be exactly one such bundle in a set of bundles for it to be a valid channel (among other things)?

And skipRange is irrelevant to channel head, or anything to do with the registry server. It's just a piece of metadata sent over to the resolver. Do I have that right?

Reply:
I updated this text to what I think is accurate.